### PR TITLE
feat: stop command support windows

### PR DESCRIPTION
**Pull request by BaffinLee**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

`stop` command supports windows now.

##### Description of change

Adjust find process command and filter logic to support windows, done simple than other pr.

Tested on Windows 7 and Windows 10, works fine.

By the way, the `npm run test` wouldn't pass even if I don't change anything, which I didn't dig in.

### CODE REVIEW

File: lib/helper.js [hunk 1]
```diff
@@ -1,18 +1,21 @@
...
 
 const runScript = require('runscript');
-const REGEX = /^\s*(\d+)\s+(.*)/;
+const isWin = process.platform === 'win32';
+const REGEX = isWin ? /^(.*)\s+(\d+)\s+$/ : /^\s*(\d+)\s+(.*)/;
 
 exports.findNodeProcess = function* (filterFn) {
-  const command = 'ps -eo "pid,command"';
+  const command = isWin ?
+    'wmic Path win32_process Where "Name = \'node.exe\'" Get CommandLine,ProcessId' :
```
Comment (popomore) [inline lib/helper.js]: I’m not familiar with Windows, is wmic a buildin command?

Comment (BaffinLee) [reply lib/helper.js]: Yes, from [Microsoft offical docs](https://docs.microsoft.com/en-us/windows/desktop/wmisdk/wmic#requirements), but I am not sure about Windows XP.

image

Comment (popomore) [reply lib/helper.js]: It seems good

Comment (atian25) [reply lib/helper.js]: image

https://support.microsoft.com/en-us/help/2664203/-invalid-xml-content-error-message-when-you-run-a-wmic-command-in-wind?sd=rss&spid=14019

Comment (BaffinLee) [reply lib/helper.js]: @atian25 It seems some people have the same issue about `Invalid XML content`. I can't figure out why, I have tested the script on 3 windows pc, works fine.

what about `wmic Path win32_process Where "Name = 'node.exe'" Get CommandLine,ProcessId`, output the same error ?

I am going to test it on `appveyor`.

Comment (atian25) [reply lib/helper.js]: worker for most of the window user is better than nothing~

**Comment by BaffinLee** [comment]
Passed every test case on [appveyor](https://ci.appveyor.com/project/BaffinLee/egg-scripts), and [travis-ci](https://travis-ci.org/eggjs/egg-scripts/builds/401310687).

image

TODO:
- [ ] find a replacement for `tail` command on windows. (not absolutly necessary)

By the way, I konw why I couldn't run test locally before, you guys use `npminstall` (`cnpm`) , but I use `npm` or `yarn`. I will fire a PR latter with detail.

File: lib/helper.js [hunk 2]
```diff
@@ -1,18 +1,21 @@
...
+    'ps -eo "pid,command"';
   const stdio = yield runScript(command, { stdio: 'pipe' });
   const processList = stdio.stdout.toString().split('\n')
     .reduce((arr, line) => {
       if (!!line && !line.includes('/bin/sh') && line.includes('node')) {
         const m = line.match(REGEX);
         /* istanbul ignore else */
         if (m) {
-          const item = { pid: m[1], cmd: m[2] };
+          const item = isWin ? { pid: m[2], cmd: m[1] } : { pid: m[1], cmd: m[2] };
```
Comment (atian25) [inline lib/helper.js]: why not just change  `Get CommandLine,ProcessId` to `Get ProcessId,CommandLine` ?

Comment (atian25) [reply lib/helper.js]: +1 after fix this

Comment (BaffinLee) [reply lib/helper.js]: In my test, `ProcessId` always show behind `CommandLine`, despite how you run the command.

So I just put `ProcessId` behind `CommandLine`, more clear, less confuse.

Comment (atian25) [reply lib/helper.js]: how about change the `ps` of linux to make they the same order ?

Comment (BaffinLee) [reply lib/helper.js]: It's still confusing, `Get CommandLine,ProcessId`  and `Get ProcessId,CommandLine` actually output the same thing, maybe I should add a comment in the code ? Got better idea? @atian25

Comment (atian25) [reply lib/helper.js]: `ps -eo "pid,command"` -> `ps -eo "command,pid"`

Comment (BaffinLee) [reply lib/helper.js]: That's better, hahaha

Comment (BaffinLee) [reply lib/helper.js]: something wrong with the `REGEXP` in my final commit, I will switch to linux to test it.

Comment (BaffinLee) [reply lib/helper.js]: It's weird:

`ps -eo pid,command` works fine

```
fake@ubuntu$ ps -eo pid,command
  PID COMMAND
    1 /init ro
   24 ssh-agent
  205 /init ro
  206 -bash
 2956 /init ro
 2957 -bash
 5274 npm
 5284 sh -c npm run lint -- --fix && npm run pkgfiles && npm run test-local
 5325 npm
 5335 sh -c egg-bin test
 5336 node /mnt/c/Users/fake/code/egg-scripts/node_modules/.bin/egg-bin test
 5342 /usr/local/lib/nodejs/node/bin/node /mnt/c/Users/fake/code/egg-scripts/node_modules/egg-bin/node_modules/mocha/bin/_mocha --require=/mnt/c/Users/fake/code/egg-scripts/node_modules/egg-bin/lib/mocha-clean.js --require=/mnt/c/U...
 5595 node /mnt/c/Users/fake/code/egg-scripts/lib/start-cluster {"framework":"/mnt/c/Users/fake/code/egg-scripts/test/fixtures/example/node_modules/yadan","workers":2,"title":"egg-server-example","baseDir":"/mnt/c/Users/fake/cod...
 5680 /usr/local/lib/nodejs/node/bin/node /mnt/c/Users/fake/code/egg-scripts/bin/egg-scripts.js start --workers=2 --title=egg-test /mnt/c/Users/fake/code/egg-scripts/test/fixtures/example
 5690 node /mnt/c/Users/fake/code/egg-scripts/lib/start-cluster {"workers":2,"title":"egg-test","framework":"/mnt/c/Users/fake/code/egg-scripts/test/fixtures/example/node_modules/custom-framework","baseDir":"/mnt/c/Users/fake/co...
 5696 /usr/local/lib/nodejs/node/bin/node /mnt/c/Users/fake/code/egg-scripts/node_modules/egg-cluster/lib/agent_worker.js {"framework":"/mnt/c/Users/fake/code/egg-scripts/test/fixtures/example/node_modules/custom-framework","baseDi...
 5706 /usr/local/lib/nodejs/node/bin/node /mnt/c/Users/fake/code/egg-scripts/node_modules/egg-cluster/lib/app_worker.js {"framework":"/mnt/c/Users/fake/code/egg-scripts/test/fixtures/example/node_modules/custom-framework","baseDir"...
 5707 /usr/local/lib/nodejs/node/bin/node /mnt/c/Users/fake/code/egg-scripts/node_modules/egg-cluster/lib/app_worker.js {"framework":"/mnt/c/Users/fake/code/egg-scripts/test/fixtures/example/node_modules/custom-framework","baseDir"...
```

`ps -eo command,pid` faild: `command` was truncated

```
fake@ubuntu$ ps -eo command,pid
COMMAND                       PID
/init ro                        1
ssh-agent                      24
/init ro                      205
-bash                         206
/init ro                     2956
-bash                        2957
npm                          5274
sh -c npm run lint -- --fix  5284
npm                          5325
sh -c egg-bin test           5335
node /mnt/c/Users/fake/c     5336
/usr/local/lib/nodejs/node/  5342
node /mnt/c/Users/fake/c     5595
/usr/local/lib/nodejs/node/  5680
node /mnt/c/Users/fake/c     5690
/usr/local/lib/nodejs/node/  5696
/usr/local/lib/nodejs/node/  5706
/usr/local/lib/nodejs/node/  5707
ps -eo command,pid           5741
```

refer to [how-to-prevent-ps-from-truncating-the-process-name](https://askubuntu.com/questions/717919/how-to-prevent-ps-from-truncating-the-process-name), I think I should rollback :) ? or set column width like: `ps -eo command:1000,pid` ?

I'd like to rollback, any idea?

Comment (BaffinLee) [reply lib/helper.js]: @atian25

Comment (atian25) [reply lib/helper.js]: no, just rollback to last review, thanks

Comment (BaffinLee) [reply lib/helper.js]: @atian25 done. Test on [appveyor](https://ci.appveyor.com/project/BaffinLee/egg-scripts) passed too.

**Comment by fengmk2** [comment]
should change README.md https://github.com/eggjs/egg-scripts/blob/master/README.md#egg-scripts too.

image

**Comment by BaffinLee** [comment]
@fengmk2 done.

File: README.md [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 deploy tool for egg project.
 
-**Note: Windows is not supported**
+**Note: Windows is partially supported, see [#22](https://github.com/eggjs/egg-scripts/pull/22)**
```
Comment (atian25) [inline README.md]: - egg guide docs of deploy is also need to updated.
- could raise anthoer PR to support `start` for win (if not `tail` instead, just ignore the checker at win)

Comment (atian25) [reply README.md]: https://ci.appveyor.com/project/eggjs/egg-scripts

Comment (BaffinLee) [reply README.md]: PR for egg's docs, [#2788](https://github.com/eggjs/egg/pull/2788)

**Comment by waitingsong** [comment]
有个现象， vscode 调试 egg 中，wmic 频繁出现在系统任务列表中。大概3秒执行一次，每次0.3秒左右。
.vscode/launch.json:
```
    {
      "name": "Egg Debug",
      "type": "node",
      "request": "launch",
      "runtimeExecutable": "npm",
      "runtimeArgs": [
        "run",
        "debug",
        "--",
        "--inspect-brk"
      ],
      "console": "integratedTerminal",
      "restart": true,
      "protocol": "auto",
      "port": 9229,
      "autoAttachChildProcesses": true
    },
```
单独运行 `npm run debug` 没发现有这状况。 这和 vscode 有关？

**Comment by BaffinLee** [comment]
是吧，调试应该没用到 `egg-scripts stop`

**Pull request merged**